File: ios/CellularChatCore/Sources/CellularChatCore/FindStateMachine.swift
import Foundation

/// Logical Find session state machine (PROTOCOL_V2.md §10). Both platforms
/// implement the identical reducer; conformance is state_transitions.json.
public enum FindState: String, CaseIterable, Equatable, Sendable {
    case idle, arming, searching, p2pConnecting, authenticating, connected
    case rangingStarting, directionAvailable, distanceOnly, proximityOnly, connectedOnly
    case signalLost, retryWait, stopped, expired, failed
}

public enum FindEvent: String, CaseIterable, Equatable, Sendable {
    case arm = "ARM"
    case armed = "ARMED"
    case peerFound = "PEER_FOUND"
    case transportConnected = "TRANSPORT_CONNECTED"
    case authenticated = "AUTHENTICATED"
    case rangingStarting = "RANGING_STARTING"
    case sampleDirection = "SAMPLE_DIRECTION"
    case sampleDistance = "SAMPLE_DISTANCE"
    case sampleProximity = "SAMPLE_PROXIMITY"
    case rangingUnavailable = "RANGING_UNAVAILABLE"
    case signalLost = "SIGNAL_LOST"
    case retryWait = "RETRY_WAIT"
    case retryElapsed = "RETRY_ELAPSED"
    case userStop = "USER_STOP"
    case deadline = "DEADLINE"
    case fatal = "FATAL"
}

public enum FindStateMachine {

    private static let terminal: Set<FindState> = [.stopped, .expired, .failed]
    private static let rangingStates: Set<FindState> = [
        .rangingStarting, .directionAvailable, .distanceOnly, .proximityOnly, .connectedOnly,
    ]

    /// Returns the next state, or nil if the (state, event) pair is invalid.
    public static func reduce(_ state: FindState, _ event: FindEvent) -> FindState? {
        // Terminal and idle states allow only their explicit transitions.
        switch (state, event) {
        case (.idle, .arm): return .arming
        case (.arming, .armed): return .searching
        case (.searching, .peerFound): return .p2pConnecting
        case (.p2pConnecting, .transportConnected): return .authenticating
        case (.authenticating, .authenticated): return .connected
        case (.connected, .rangingStarting): return .rangingStarting
        case (.connectedOnly, .rangingStarting): return .rangingStarting
        case (.signalLost, .retryWait): return .retryWait
        case (.retryWait, .retryElapsed): return .searching
        default:
            break
        }

        // RANGING_UNAVAILABLE from connected / ranging (non-connectedOnly) -> connectedOnly.
        if event == .rangingUnavailable,
           state == .connected || (rangingStates.contains(state) && state != .connectedOnly) {
            return .connectedOnly
        }

        // Sample events from ranging states (excluding connectedOnly).
        let sampleSources: Set<FindState> = [
            .rangingStarting, .directionAvailable, .distanceOnly, .proximityOnly,
        ]
        if sampleSources.contains(state) {
            switch event {
            case .sampleDirection: return .directionAvailable
            case .sampleDistance: return .distanceOnly
            case .sampleProximity: return .proximityOnly
            default: break
            }
        }

        // SIGNAL_LOST from connected | ranging | p2pConnecting | authenticating.
        if event == .signalLost {
            let linked: Set<FindState> = rangingStates.union([.connected, .p2pConnecting, .authenticating])
            if linked.contains(state) { return .signalLost }
        }

        // USER_STOP / DEADLINE / FATAL from any non-terminal, non-idle state.
        if state != .idle && !terminal.contains(state) {
            switch event {
            case .userStop: return .stopped
            case .deadline: return .expired
            case .fatal: return .failed
            default: break
            }
        }

        return nil
    }
}

File: android/app/src/main/java/com/cellularchat/app/core/protocol/FindStateMachine.kt
package com.cellularchat.app.core.protocol

/** Logical Find session states (PROTOCOL_V2.md §10). [wire] matches the fixtures. */
enum class FindState(val wire: String) {
    IDLE("idle"),
    ARMING("arming"),
    SEARCHING("searching"),
    P2P_CONNECTING("p2pConnecting"),
    AUTHENTICATING("authenticating"),
    CONNECTED("connected"),
    RANGING_STARTING("rangingStarting"),
    DIRECTION_AVAILABLE("directionAvailable"),
    DISTANCE_ONLY("distanceOnly"),
    PROXIMITY_ONLY("proximityOnly"),
    CONNECTED_ONLY("connectedOnly"),
    SIGNAL_LOST("signalLost"),
    RETRY_WAIT("retryWait"),
    STOPPED("stopped"),
    EXPIRED("expired"),
    FAILED("failed");

    companion object {
        fun fromWire(wire: String): FindState =
            entries.first { it.wire == wire }
    }
}

enum class FindEvent {
    ARM, ARMED, PEER_FOUND, TRANSPORT_CONNECTED, AUTHENTICATED, RANGING_STARTING,
    SAMPLE_DIRECTION, SAMPLE_DISTANCE, SAMPLE_PROXIMITY, RANGING_UNAVAILABLE,
    SIGNAL_LOST, RETRY_WAIT, RETRY_ELAPSED, USER_STOP, DEADLINE, FATAL;

    companion object {
        fun fromWire(wire: String): FindEvent = valueOf(wire)
    }
}

data class FindTransition(val next: FindState, val valid: Boolean)

/** Deterministic Find reducer; conformance fixtures in state_transitions.json. */
object FindStateMachine {
    fun reduce(state: FindState, event: FindEvent): FindTransition {
        val next = when (state) {
            FindState.IDLE -> when (event) {
                FindEvent.ARM -> FindState.ARMING
                else -> null
            }
            FindState.ARMING -> when (event) {
                FindEvent.ARMED -> FindState.SEARCHING
                else -> terminal(event)
            }
            FindState.SEARCHING -> when (event) {
                FindEvent.PEER_FOUND -> FindState.P2P_CONNECTING
                else -> terminal(event)
            }
            FindState.P2P_CONNECTING -> when (event) {
                FindEvent.TRANSPORT_CONNECTED -> FindState.AUTHENTICATING
                FindEvent.SIGNAL_LOST -> FindState.SIGNAL_LOST
                else -> terminal(event)
            }
            FindState.AUTHENTICATING -> when (event) {
                FindEvent.AUTHENTICATED -> FindState.CONNECTED
                FindEvent.SIGNAL_LOST -> FindState.SIGNAL_LOST
                else -> terminal(event)
            }
            FindState.CONNECTED -> when (event) {
                FindEvent.RANGING_STARTING -> FindState.RANGING_STARTING
                FindEvent.RANGING_UNAVAILABLE -> FindState.CONNECTED_ONLY
                FindEvent.SIGNAL_LOST -> FindState.SIGNAL_LOST
                else -> terminal(event)
            }
            FindState.RANGING_STARTING,
            FindState.DIRECTION_AVAILABLE,
            FindState.DISTANCE_ONLY,
            FindState.PROXIMITY_ONLY -> when (event) {
                FindEvent.SAMPLE_DIRECTION -> FindState.DIRECTION_AVAILABLE
                FindEvent.SAMPLE_DISTANCE -> FindState.DISTANCE_ONLY
                FindEvent.SAMPLE_PROXIMITY -> FindState.PROXIMITY_ONLY
                FindEvent.RANGING_UNAVAILABLE -> FindState.CONNECTED_ONLY
                FindEvent.SIGNAL_LOST -> FindState.SIGNAL_LOST
                else -> terminal(event)
            }
            FindState.CONNECTED_ONLY -> when (event) {
                FindEvent.RANGING_STARTING -> FindState.RANGING_STARTING
                FindEvent.SIGNAL_LOST -> FindState.SIGNAL_LOST
                else -> terminal(event)
            }
            FindState.SIGNAL_LOST -> when (event) {
                FindEvent.RETRY_WAIT -> FindState.RETRY_WAIT
                else -> terminal(event)
            }
            FindState.RETRY_WAIT -> when (event) {
                FindEvent.RETRY_ELAPSED -> FindState.SEARCHING
                else -> terminal(event)
            }
            FindState.STOPPED, FindState.EXPIRED, FindState.FAILED -> null
        }
        return if (next == null) FindTransition(state, false) else FindTransition(next, true)
    }

    /** USER_STOP/DEADLINE/FATAL are accepted from every non-terminal active state. */
    private fun terminal(event: FindEvent): FindState? = when (event) {
        FindEvent.USER_STOP -> FindState.STOPPED
        FindEvent.DEADLINE -> FindState.EXPIRED
        FindEvent.FATAL -> FindState.FAILED
        else -> null
    }
}
